cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -0,0 +1,351 @@
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.cdap.security.spi.authorization.UnauthorizedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Class that applies capabilities
+ */
+public class CapabilityApplier {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CapabilityApplier.class);
+  private static final Gson GSON = new Gson();
+  private static final int RETRY_LIMIT = 5;
+  private static final int RETRY_DELAY = 5;
+  private static final ProgramTerminator NOOP_PROGRAM_TERMINATOR = programId -> {
+    // no-op
+  };
+  private final SystemProgramManagementService systemProgramManagementService;
+  private final ApplicationLifecycleService applicationLifecycleService;
+  private final ProgramLifecycleService programLifecycleService;
+  private final NamespaceAdmin namespaceAdmin;
+  private final CapabilityReader capabilityReader;
+  private final CapabilityWriter capabilityWriter;
+
+  @Inject
+  CapabilityApplier(CConfiguration cConf, SystemProgramManagementService systemProgramManagementService,
+                    ApplicationLifecycleService applicationLifecycleService, NamespaceAdmin namespaceAdmin,
+                    ProgramLifecycleService programLifecycleService, CapabilityReader capabilityChecker,
+                    CapabilityWriter capabilityWriter) {
+    this.systemProgramManagementService = systemProgramManagementService;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.programLifecycleService = programLifecycleService;
+    this.capabilityReader = capabilityChecker;
+    this.capabilityWriter = capabilityWriter;
+    this.namespaceAdmin = namespaceAdmin;
+  }
+
+  /**
+   * Applies the given capability configurations
+   *
+   * @param capabilityConfigs
+   */
+  public void apply(List<CapabilityConfig> capabilityConfigs) throws Exception {
+    List<CapabilityConfig> newConfigs = new ArrayList<>(capabilityConfigs);
+    Set<CapabilityConfig> enableSet = new HashSet<>();
+    Set<CapabilityConfig> disableSet = new HashSet<>();
+    Set<CapabilityConfig> deleteSet = new HashSet<>();
+    Map<String, CapabilityStatusRecord> currentCapabilities = capabilityReader.getAllCapabilities().stream().collect(
+      Collectors.toMap(CapabilityStatusRecord::getCapability, capabilityStatusRecord -> capabilityStatusRecord));
+    Map<String, CapabilityOperationRecord> currentOperations = capabilityReader.getCapabilityOperations().stream()
+      .collect(Collectors.toMap(CapabilityOperationRecord::getCapability,
+                                capabilityOperationRecord -> capabilityOperationRecord));
```

Comment:
for identity function, use `Function.identity()` instead

Reply:
Done